odl/space/gpuary_ntuples.py

```diff
@@ -0,0 +1,2302 @@
+                         '`out.shape` {oshp}'
+                         ''.format(result_shape, np.shape(a), x.shape,
+                                   np.shape(b), y.shape, oshp=out.shape))
+
+    a_arg = pygpu.elemwise.as_argument(a, 'a', read=True)
+    x_arg = pygpu.elemwise.as_argument(x, 'x', read=True)
+    b_arg = pygpu.elemwise.as_argument(b, 'b', read=True)
+    y_arg = pygpu.elemwise.as_argument(y, 'y', read=True)
+    out_arg = pygpu.elemwise.as_argument(out, 'out', write=True)
+    args = [a_arg, x_arg, b_arg, y_arg, out_arg]
+
+    oper = 'out = a * x + b * y'
+    # TODO: check what to do with `convert_f16`
+    kernel = pygpu.elemwise.GpuElemwise(out.context, oper, args,
+                                        convert_f16=True)
+    kernel(a, x, b, y, out, broadcast=True)
+
+
+def _lincomb(a, x1, b, x2, out):
+    """Raw linear combination, assuming types have been checked."""
+
+    if x1 is x2 and b != 0:
+        # x1 is aligned with x2 =>  out <-- (a + b) * x1
+        scal(a + b, x1.data, out.data)
+    elif out is x1 and out is x2:
+        # All the vectors are aligned =>  out <-- (a + b) * out
+        out.data *= a + b
+    elif out is x1:
+        # out is aligned with x1 =>  out <-- a * out + b * x2
+        axpby(b, x2.data, a, out.data)
+    elif out is x2:
+        # out is aligned with x2 =>  out <-- a * x1 + b * out
+        if a == 0:
+            # Use in-place  out <-- b * out directly
+            out.data *= b
+        elif b == 0:  # out <-- a * x1
+            scal(a, x1.data, out.data)
+        else:  # out <-- a * x1 + b * out
+            axpby(a, x1.data, b, out.data)
+    else:
+        # We have exhausted all alignment options, so x1 != x2 != out
+        # We now optimize for various values of a and b
+        if b == 0:
+            if a == 0:  # Zero assignment  out <-- 0
+                out.data[:] = 0
+            elif a == 1:  # Copy  out <-- x1
+                out.data[:] = x1.data
+            else:  # Scaled copy  out <-- a * x1
+                axpy(a, x1.data, out.data)
+        else:
+            if a == 0:
+                if b == 1:
+                    # Copy  out <-- x2
+                    out.data[:] = x2.data
+                else:
+                    # Scaled copy  out <-- b * x2
+                    scal(b, x2.data, out.data)
+            elif a == 1:  # No scaling in x1 =>  out <-- x1 + b * x2
+                # TODO: make optimized version or change axpy
+                lico(1, x1.data, b, x2.data, out.data)
+            else:
+                if b == 1:  # No scaling in x2 =>  out <-- a * x1 + x2
+                    # TODO: make optimized version or change axpy
+                    lico(a, x1.data, 1, x2.data, out.data)
+                else:  # General case  out <-- a * x1 + b * x2
+                    lico(a, x1.data, b, x2.data, out.data)
+
+
+class GpuArrayTensorSpace(FnBase, GpuArrayRawTensorSpace):
```

Comment:
From here on, the rest is currently copy-paste from `npy_ntuples.py`. Ignore it.